README.md
# Random string generator for PHP

[![Latest Version on Packagist](https://img.shields.io/packagist/v/stfn/php-random-string.svg?style=flat-square)](https://packagist.org/packages/stfn/php-random-string)
[![Tests](https://img.shields.io/github/actions/workflow/status/stfndamjanovic/php-random-string/run-tests.yml?branch=main&label=tests&style=flat-square)](https://github.com/stfndamjanovic/php-random-string/actions/workflows/run-tests.yml)

This package can be used to generate a random string based on your set of characters or predefined ones. You can configure string length, prefix, suffix, and count of strings, or skip some strings under certain conditions...

## Installation

You can install the package via composer:

```bash
composer require stfn/php-random-string
```

## Usage

### Basic
Simple example without any configuration.

```php
$string = RandomString::new()->generate(); // Output: RIKdjFzuDaN12RiJ
```

### Length definition

You can control the length of the string. By default, it's 16 characters.

```php
$string = RandomString::new(6)->generate(); // Output: dzGcot
````

### Predefined charset

If you want to generate a string consisting of numbers only, lowercase letters, or uppercase letters you can use predefined charsets.

```php
// Generate string that contains only numbers
$config = StringConfig::make()
            ->numbersOnly();

$string = RandomString::fromConfig($config)->generate(); // Output: 9387406871490781

// Generate string that contains only lowercase letters
$config = StringConfig::make()
            ->lowerCaseOnly();

$string = RandomString::fromConfig($config)->generate(); // Output: hvphyfmgnvbbajve

// Generate string that contains only uppercase letters
$config = StringConfig::make()
            ->upperCaseOnly();

$string = RandomString::fromConfig($config)->generate(); // Output: ZIVSUDQHAMDNQAYV
```

### Custom charset

Or you can use your custom charset for generating random string.

```php
$config = StringConfig::make()
            ->charset("ABCDEFG1234");

$string = RandomString::fromConfig($config)->generate(); // Output: 3B41B32C2A12A3A1
```

### Skipping

Sometimes you may want to generate a random string but under certain conditions. 
For example, give me a string that is not part of this array.

```php
$config = StringConfig::make()
            ->numbersOnly()
            ->length(6)
            ->skip(function ($string) {
                return in_array($string, ["025922", "104923"]);
            });

$string = RandomString::fromConfig($config)->generate(); // Output: 083712
```

### Prefix and Suffix

If you want to add a prefix or suffix to generated string, you can do it like this.

```php
$config = StringConfig::make()
            ->length(6)
            ->prefix("PRE_")
            ->suffix("_AFTER");

$string = RandomString::fromConfig($config)->generate(); // Output: PRE_rkM7Jl_AFTER
```

### Array of random strings

`RandomString` can generate more than just one string.

```php
$config = StringConfig::make()
            ->length(6)
            ->count(3);

$strings = RandomString::fromConfig($config)->generate();

// Output: ["ozBYeT", "BYjCtr", "Sw7O5b"];
```

### Uniqueness

It may happen (rarely, but it's possible) to have not unique strings in the generated array. If you want to avoid it, just change the config.

```php
$config = StringConfig::make()
            ->length(6)
            ->count(3)
            ->unique();

$strings = RandomString::fromConfig($config)->generate();

// Output: ["92ONRj", "Me6oym", "WbBPVc"];
```

### Everything in one line

You can use the `fromArray` method if you don't want to create 2 objects every time.

```php
$string = RandomString::fromArray(['length' => 6, 'charset' => 'ABCD1234'])->generate(); // Output: CCDA1D
```

## Testing

```bash
composer test
```

## Security

While the `RandomString` class is designed to generate random and unpredictable string, it is important to note that it is not a cryptographically secure hash function and should not be used for sensitive applications such as password hashing or cryptographic key generation.

## Changelog

Please see [CHANGELOG](CHANGELOG.md) for more information on what has changed recently.

## Credits

- [Stefan Damjanovic](https://github.com/stfndamjanovic)
- [All Contributors](../../contributors)

## License

The MIT License (MIT). Please see [License File](LICENSE.md) for more information.


CHANGELOG.md
# Changelog

All notable changes to `php-random-string` will be documented in this file.

## 1.0.0 - 2023-03-28

Stable release


tests/RandomHashTest.php
<?php

namespace Stfn\RandomHash\Tests;

use PHPUnit\Framework\TestCase;
use Stfn\RandomHash\HashConfig;
use Stfn\RandomHash\InvalidConfigException;
use Stfn\RandomHash\RandomHash;

class RandomHashTest extends TestCase
{
    public function test_if_it_can_generate_hash_with_specific_length()
    {
        $config = new HashConfig();
        $config->length(10);

        $instance = new RandomHash($config);
        $hash = $instance->generate();

        $this->assertEquals(10, strlen($hash));

        $config->length(12);
        $hash = $instance->generate();

        $this->assertEquals(12, strlen($hash));
    }

    public function test_if_it_can_generate_hash_with_specific_prefix()
    {
        $config = new HashConfig();
        $config->prefix('PREFIX');

        $instance = new RandomHash($config);
        $hash = $instance->generate();

        $this->assertStringStartsWith('PREFIX', $hash);

        $config->prefix('NEW_PREFIX');
        $hash = $instance->generate();

        $this->assertStringStartsWith('NEW_PREFIX', $hash);
    }

    public function test_if_it_can_generate_hash_with_specific_suffix()
    {
        $config = new HashConfig();
        $config->suffix('SUFFIX');

        $instance = new RandomHash($config);
        $hash = $instance->generate();

        $this->assertStringEndsWith('SUFFIX', $hash);

        $config->suffix('NEW_SUFFIX');
        $hash = $instance->generate();

        $this->assertStringEndsWith('NEW_SUFFIX', $hash);
    }

    public function test_if_it_can_generate_hash_with_specific_count()
    {
        $config = new HashConfig();
        $config->count(10);

        $instance = new RandomHash($config);
        $hash = $instance->generate();

        $this->assertCount(10, $hash);
    }

    public function test_if_it_can_generate_only_numbers()
    {
        $config = new HashConfig();
        $config->length(6)
            ->numbersOnly();

        $instance = new RandomHash($config);
        $hash = $instance->generate();

        $this->assertMatchesRegularExpression('/^\d{6}$/', $hash);

        $config->length(8);
        $hash = $instance->generate();

        $this->assertMatchesRegularExpression('/^\d{8}$/', $hash);
    }

    public function test_if_it_can_generate_uppercase_letters()
    {
        $config = new HashConfig();
        $config->length(6)
            ->upperCaseOnly();

        $instance = new RandomHash($config);
        $hash = $instance->generate();

        $this->assertMatchesRegularExpression('/^[A-Z]{6}$/', $hash);

        $config->length(8);
        $hash = $instance->generate();

        $this->assertMatchesRegularExpression('/^[A-Z]{8}$/', $hash);
    }

    public function test_if_it_can_generate_lowercase_letters()
    {
        $config = new HashConfig();
        $config->length(6)
            ->lowerCaseOnly();

        $instance = new RandomHash($config);
        $hash = $instance->generate();

        $this->assertMatchesRegularExpression('/^[a-z]{6}$/', $hash);

        $config->length(8);
        $hash = $instance->generate();

        $this->assertMatchesRegularExpression('/^[a-z]{8}$/', $hash);
    }

    public function test_if_it_can_generate_alphanumeric()
    {
        $config = new HashConfig();
        $config->length(10)
            ->charset(HashConfig::CHARSET_UPPERCASE.HashConfig::CHARSET_LOWERCASE.HashConfig::CHARSET_NUMERIC);

        $instance = new RandomHash($config);
        $hash = $instance->generate();

        $this->assertMatchesRegularExpression('/^[0-9a-zA-Z]{10}$/', $hash);

        $config->length(12);
        $hash = $instance->generate();

        $this->assertMatchesRegularExpression('/^[0-9a-zA-Z]{12}$/', $hash);
    }

    public function test_if_it_can_use_skip_callback()
    {
        $config = new HashConfig();
        $config->length(2)
            ->charset('01')
            ->skip(function ($hash) {
                return in_array($hash, ['10', '11', '01']);
            });

        $instance = new RandomHash($config);
        $hash = $instance->generate();

        $this->assertEquals('00', $hash);
    }

    public function test_if_it_can_generate_not_unique_hashes()
    {
        $config = new HashConfig();
        $config->length(2)
            ->count(5)
            ->notUnique()
            ->charset('01')
            ->skip(function ($hash) {
                return in_array($hash, ['10', '11', '01']);
            });

        $instance = new RandomHash($config);
        $hashes = $instance->generate();

        foreach ($hashes as $hash) {
            $this->assertEquals('00', $hash);
        }
    }

    public function test_if_it_can_generate_unique_hashes()
    {
        $config = new HashConfig();
        $config->length(2)
            ->count(2)
            ->unique()
            ->charset('01')
            ->skip(function ($hash) {
                return in_array($hash, ['10', '11']);
            });

        $instance = new RandomHash($config);
        $hashes = $instance->generate();

        $this->assertEqualsCanonicalizing(['01', '00'], $hashes);
    }

    public function test_if_it_fails_after_max_possible_combination_reached()
    {
        $config = new HashConfig();
        $config->length(2)
            ->charset('01')
            ->skip(function ($hash) {
                return in_array($hash, ['10', '11', '01', '00']);
            });

        $instance = new RandomHash($config);

        $this->expectException(InvalidConfigException::class);
        $instance->generate();
    }

    public function test_if_it_fails_after_max_possible_combination_reached_using_unique()
    {
        $config = new HashConfig();
        $config->length(2)
            ->charset('01')
            ->count(3)
            ->unique()
            ->skip(function ($hash) {
                return in_array($hash, ['10', '11']);
            });

        $instance = new RandomHash($config);

        $this->expectException(InvalidConfigException::class);
        $instance->generate();
    }

    public function test_if_it_can_generate_hashes_from_array_config()
    {
        $prefix = 'TEST_';

        $hashes = RandomHash::fromConfig([
            'length' => 12,
            'prefix' => $prefix,
            'count' => 10,
        ])->generate();

        $this->assertStringStartsWith($prefix, $hashes[0]);
        $this->assertEquals(12 + strlen($prefix), strlen($hashes[0]));
        $this->assertCount(10, $hashes);
    }

    public function test_if_it_can_generate_hash_from_config_object()
    {
        $config = HashConfig::fromArray(['length' => 8]);

        $hash = RandomHash::make($config)->generate();

        $this->assertEquals(8, strlen($hash));
    }

    public function test_if_it_can_generate_a_lot_of_not_unique_hashes()
    {
        $config = new HashConfig();

        $config->length(10)
            ->count(1000000);

        $hashes = RandomHash::make($config)->generate();

        $this->assertCount(1000000, $hashes);
    }

    public function test_if_negative_number_cannot_be_used_for_count()
    {
        $config = new HashConfig();
        $config->count(-1);

        $this->expectException(InvalidConfigException::class);
        RandomHash::make($config)->generate();
    }

    public function test_if_negative_number_cannot_be_used_for_length()
    {
        $config = new HashConfig();
        $config->length(-1);

        $this->expectException(InvalidConfigException::class);
        RandomHash::make($config)->generate();
    }

    public function test_if_charset_cannot_be_empty_string()
    {
        $config = new HashConfig();
        $config->charset('');

        $this->expectException(InvalidConfigException::class);
        RandomHash::make($config)->generate();
    }
}


tests/RandomStringTest.php
<?php

namespace Stfn\RandomString\Tests;

use PHPUnit\Framework\TestCase;
use Stfn\RandomString\InvalidStringConfigException;
use Stfn\RandomString\RandomString;
use Stfn\RandomString\StringConfig;

class RandomStringTest extends TestCase
{
    public function test_if_it_can_generate_string_without_configuration()
    {
        $string = RandomString::new()->generate();

        $this->assertEquals(16, strlen($string));
    }

    public function test_if_it_can_generate_string_with_specific_length()
    {
        $config = new StringConfig();
        $config->length(10);

        $instance = new RandomString($config);
        $string = $instance->generate();

        $this->assertEquals(10, strlen($string));

        $config->length(12);
        $string = $instance->generate();

        $this->assertEquals(12, strlen($string));
    }

    public function test_if_it_can_generate_string_with_specific_prefix()
    {
        $config = new StringConfig();
        $config->prefix('PREFIX');

        $instance = new RandomString($config);
        $string = $instance->generate();

        $this->assertStringStartsWith('PREFIX', $string);

        $config->prefix('NEW_PREFIX');
        $string = $instance->generate();

        $this->assertStringStartsWith('NEW_PREFIX', $string);
    }

    public function test_if_it_can_generate_string_with_specific_suffix()
    {
        $config = new StringConfig();
        $config->suffix('SUFFIX');

        $instance = new RandomString($config);
        $string = $instance->generate();

        $this->assertStringEndsWith('SUFFIX', $string);

        $config->suffix('NEW_SUFFIX');
        $string = $instance->generate();

        $this->assertStringEndsWith('NEW_SUFFIX', $string);
    }

    public function test_if_it_can_generate_string_with_specific_count()
    {
        $config = new StringConfig();
        $config->count(10);

        $instance = new RandomString($config);
        $string = $instance->generate();

        $this->assertCount(10, $string);
    }

    public function test_if_it_can_generate_only_numbers()
    {
        $config = new StringConfig();
        $config->length(6)
            ->numbersOnly();

        $instance = new RandomString($config);
        $string = $instance->generate();

        $this->assertMatchesRegularExpression('/^\d{6}$/', $string);

        $config->length(8);
        $string = $instance->generate();

        $this->assertMatchesRegularExpression('/^\d{8}$/', $string);
    }

    public function test_if_it_can_generate_uppercase_letters()
    {
        $config = new StringConfig();
        $config->length(6)
            ->upperCaseOnly();

        $instance = new RandomString($config);
        $string = $instance->generate();

        $this->assertMatchesRegularExpression('/^[A-Z]{6}$/', $string);

        $config->length(8);
        $string = $instance->generate();

        $this->assertMatchesRegularExpression('/^[A-Z]{8}$/', $string);
    }

    public function test_if_it_can_generate_lowercase_letters()
    {
        $config = new StringConfig();
        $config->length(6)
            ->lowerCaseOnly();

        $instance = new RandomString($config);
        $string = $instance->generate();

        $this->assertMatchesRegularExpression('/^[a-z]{6}$/', $string);

        $config->length(8);
        $string = $instance->generate();

        $this->assertMatchesRegularExpression('/^[a-z]{8}$/', $string);
    }

    public function test_if_it_can_generate_alphanumeric()
    {
        $config = new StringConfig();
        $config->length(10)
            ->charset(StringConfig::CHARSET_UPPERCASE.StringConfig::CHARSET_LOWERCASE.StringConfig::CHARSET_NUMERIC);

        $instance = new RandomString($config);
        $string = $instance->generate();

        $this->assertMatchesRegularExpression('/^[0-9a-zA-Z]{10}$/', $string);

        $config->length(12);
        $string = $instance->generate();

        $this->assertMatchesRegularExpression('/^[0-9a-zA-Z]{12}$/', $string);
    }

    public function test_if_it_can_use_skip_callback()
    {
        $config = new StringConfig();
        $config->length(2)
            ->charset('01')
            ->skip(function ($string) {
                return in_array($string, ['10', '11', '01']);
            });

        $instance = new RandomString($config);
        $string = $instance->generate();

        $this->assertEquals('00', $string);
    }

    public function test_if_it_can_generate_not_unique_strings()
    {
        $config = new StringConfig();
        $config->length(2)
            ->count(5)
            ->notUnique()
            ->charset('01')
            ->skip(function ($string) {
                return in_array($string, ['10', '11', '01']);
            });

        $instance = new RandomString($config);
        $strings = $instance->generate();

        foreach ($strings as $string) {
            $this->assertEquals('00', $string);
        }
    }

    public function test_if_it_can_generate_unique_strings()
    {
        $config = new StringConfig();
        $config->length(2)
            ->count(2)
            ->unique()
            ->charset('01')
            ->skip(function ($string) {
                return in_array($string, ['10', '11']);
            });

        $instance = new RandomString($config);
        $strings = $instance->generate();

        $this->assertEqualsCanonicalizing(['01', '00'], $strings);
    }

    public function test_if_it_fails_after_max_possible_combination_reached()
    {
        $config = new StringConfig();
        $config->length(2)
            ->charset('01')
            ->skip(function ($string) {
                return in_array($string, ['10', '11', '01', '00']);
            });

        $instance = new RandomString($config);

        $this->expectException(InvalidStringConfigException::class);
        $instance->generate();
    }

    public function test_if_it_fails_when_count_is_greater_than_possible_combinations()
    {
        $config = new StringConfig();
        $config->length(2)
            ->count(5)
            ->unique()
            ->charset('01');

        $instance = new RandomString($config);

        $this->expectException(InvalidStringConfigException::class);
        $instance->generate();
    }

    public function test_if_it_fails_after_max_possible_combination_reached_using_unique()
    {
        $config = new StringConfig();
        $config->length(2)
            ->charset('01')
            ->count(3)
            ->unique()
            ->skip(function ($string) {
                return in_array($string, ['10', '11']);
            });

        $instance = new RandomString($config);

        $this->expectException(InvalidStringConfigException::class);
        $instance->generate();
    }

    public function test_if_it_can_generate_strings_from_array_config()
    {
        $prefix = 'TEST_';

        $strings = RandomString::fromArray([
            'length' => 12,
            'prefix' => $prefix,
            'count' => 10,
        ])->generate();

        $this->assertStringStartsWith($prefix, $strings[0]);
        $this->assertEquals(12 + strlen($prefix), strlen($strings[0]));
        $this->assertCount(10, $strings);
    }

    public function test_if_it_can_generate_string_from_config_object()
    {
        $config = StringConfig::make(8);

        $string = RandomString::fromConfig($config)->generate();

        $this->assertEquals(8, strlen($string));
    }

    public function test_if_it_can_generate_a_lot_of_not_unique_strings()
    {
        $config = new StringConfig();

        $config->length(10)
            ->count(1000000);

        $strings = RandomString::fromConfig($config)->generate();

        $this->assertCount(1000000, $strings);
    }

    public function test_if_it_can_change_config_on_fly()
    {
        $config = StringConfig::make(12);
        $instance = new RandomString($config);

        $newConfig = StringConfig::make(14);
        $instance->useConfig($newConfig);

        $this->assertNotEquals($config, $instance->getConfig());
    }

    public function test_if_negative_number_cannot_be_used_for_count()
    {
        $config = new StringConfig();
        $config->count(-1);

        $this->expectException(InvalidStringConfigException::class);
        RandomString::fromConfig($config)->generate();
    }

    public function test_if_negative_number_cannot_be_used_for_length()
    {
        $config = new StringConfig();
        $config->length(-1);

        $this->expectException(InvalidStringConfigException::class);
        RandomString::fromConfig($config)->generate();
    }

    public function test_if_charset_cannot_be_empty_string()
    {
        $config = new StringConfig();
        $config->charset('');

        $this->expectException(InvalidStringConfigException::class);
        RandomString::fromConfig($config)->generate();
    }
}


src/HashConfig.php
<?php

declare(strict_types=1);

namespace Stfn\RandomHash;

class HashConfig
{
    const CHARSET_UPPERCASE = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ';

    const CHARSET_LOWERCASE = 'abcdefghijklmnopqrstuvwxyz';

    const CHARSET_NUMERIC = '0123456789';

    protected $length;

    protected $count = 1;

    protected $charset;

    protected $prefix = '';

    protected $suffix = '';

    protected $unique = false;

    protected $skipCallback;

    public function __construct($length = 16)
    {
        $this->length = $length;
        $this->charset = self::CHARSET_LOWERCASE.self::CHARSET_UPPERCASE.self::CHARSET_NUMERIC;
    }

    public static function make($length = 16)
    {
        return new self($length);
    }

    public static function fromArray(array $array)
    {
        $object = new self();

        $allowedProperties = ['count', 'prefix', 'suffix', 'length', 'charset', 'unique'];

        foreach ($array as $property => $value) {
            if (property_exists($object, $property) && in_array($property, $allowedProperties)) {
                $object->{$property} = $value;
            }
        }

        return $object;
    }

    public function validate()
    {
        if (! is_string($this->charset) || empty($this->charset)) {
            throw InvalidConfigException::invalidCharset();
        }

        if ($this->length < 1) {
            throw InvalidConfigException::propertyNotPositiveNumber('length');
        }

        if ($this->count < 1) {
            throw InvalidConfigException::propertyNotPositiveNumber('count');
        }
    }

    public function charset(string $charset): self
    {
        $this->charset = $charset;

        return $this;
    }

    public function upperCaseOnly(): self
    {
        $this->charset = self::CHARSET_UPPERCASE;

        return $this;
    }

    public function lowerCaseOnly(): self
    {
        $this->charset = self::CHARSET_LOWERCASE;

        return $this;
    }

    public function numbersOnly(): self
    {
        $this->charset = self::CHARSET_NUMERIC;

        return $this;
    }

    public function prefix(string $prefix)
    {
        $this->prefix = $prefix;

        return $this;
    }

    public function suffix(string $suffix)
    {
        $this->suffix = $suffix;

        return $this;
    }

    public function length(int $length): self
    {
        $this->length = $length;

        return $this;
    }

    public function skip(callable $callback)
    {
        $this->skipCallback = $callback;

        return $this;
    }

    public function unique()
    {
        $this->unique = true;

        return $this;
    }

    public function notUnique()
    {
        $this->unique = false;

        return $this;
    }

    public function hasSkipCallback()
    {
        return is_callable($this->skipCallback);
    }

    public function getSkipCallback()
    {
        return $this->skipCallback;
    }

    public function getLength(): int
    {
        return $this->length;
    }

    public function getCharset()
    {
        return $this->charset;
    }

    public function getPrefix()
    {
        return $this->prefix;
    }

    public function getSuffix()
    {
        return $this->suffix;
    }

    public function isUnique()
    {
        return $this->unique;
    }

    public function count(int $count)
    {
        $this->count = $count;

        return $this;
    }

    public function getCount()
    {
        return $this->count;
    }
}


src/RandomHash.php
<?php

declare(strict_types=1);

namespace Stfn\RandomHash;

class RandomHash
{
    protected HashConfig $config;

    protected array $hashes = [];

    private array $skipped = [];

    public function __construct(HashConfig $hashConfig)
    {
        $this->config = $hashConfig;
    }

    public static function make(HashConfig $hashConfig)
    {
        return new self($hashConfig);
    }

    public static function fromConfig($array)
    {
        return new self(HashConfig::fromArray($array));
    }

    public function generate()
    {
        $this->config->validate();

        $this->reset();

        $charset = $this->config->getCharset();
        $length = $this->config->getLength();

        $charsetLength = strlen($this->config->getCharset());
        $maxCombinations = pow($charsetLength, $this->config->getLength());

        for ($i = 0; $i < $this->config->getCount(); $i++) {
            if ($maxCombinations <= count($this->skipped)) {
                throw InvalidConfigException::maxCombinationReached();
            }

            $randomBytes = random_bytes($length);

            for ($j = $length - 1; $j > 0; $j--) {
                $rand = random_int(0, $j);
                [$randomBytes[$j], $randomBytes[$rand]] = [$randomBytes[$rand], $randomBytes[$j]];
            }

            $hash = '';

            for ($j = 0; $j < $length; $j++) {
                $hash .= $charset[ord($randomBytes[$j]) % $charsetLength];
            }

            $hash = $this->config->getPrefix().$hash.$this->config->getSuffix();

            if ($this->shouldSkip($hash)) {
                $i--;
                $this->skipHash($hash);

                continue;
            }

            $this->addHash($hash);
        }

        if (count($this->hashes) == 0) {
            return null;
        }

        if (count($this->hashes) == 1) {
            return reset($this->hashes);
        }

        return $this->hashes;
    }

    public function useConfig(HashConfig $hashConfig)
    {
        $this->config = $hashConfig;
    }

    protected function addHash(string $hash)
    {
        $this->hashes[] = $hash;
    }

    protected function getHashes()
    {
        return $this->hashes;
    }

    protected function shouldSkip($hash)
    {
        $shouldSkip = false;

        if ($this->config->isUnique()) {
            $shouldSkip = in_array($hash, $this->getHashes());
        }

        if ($this->config->hasSkipCallback()) {
            $shouldSkip = $shouldSkip || $this->config->getSkipCallback()($hash);
        }

        return $shouldSkip;
    }

    protected function reset()
    {
        $this->hashes = [];
        $this->skipped = [];
    }

    protected function skipHash($hash)
    {
        if (! in_array($hash, $this->skipped)) {
            $this->skipped[] = $hash;
        }
    }
}


src/RandomString.php
<?php

declare(strict_types=1);

namespace Stfn\RandomString;

class RandomString
{
    protected StringConfig $config;

    protected array $strings = [];

    protected array $skipped = [];

    public function __construct(StringConfig $config)
    {
        $this->config = $config;
    }

    public static function new($length = 16): self
    {
        return new self(new StringConfig($length));
    }

    public static function fromConfig(StringConfig $config): self
    {
        return new self($config);
    }

    public static function fromArray(array $input): self
    {
        return new self(StringConfig::fromArray($input));
    }

    public function generate(): string|array
    {
        $this->config->validate();

        $this->reset();

        $charset = $this->config->getCharset();
        $length = $this->config->getLength();

        $charsetLength = strlen($this->config->getCharset());
        $maxCombinations = pow($charsetLength, $this->config->getLength());

        for ($i = 0; $i < $this->config->getCount(); $i++) {
            if ($maxCombinations <= count($this->skipped)) {
                throw InvalidStringConfigException::maxCombinationReached();
            }

            $randomBytes = random_bytes($length);

            for ($j = $length - 1; $j > 0; $j--) {
                $randomInt = random_int(0, $j);
                [$randomBytes[$j], $randomBytes[$randomInt]] = [$randomBytes[$randomInt], $randomBytes[$j]];
            }

            $string = '';

            for ($j = 0; $j < $length; $j++) {
                $string .= $charset[ord($randomBytes[$j]) % $charsetLength];
            }

            $string = $this->config->getPrefix().$string.$this->config->getSuffix();

            if ($this->shouldSkip($string)) {
                $i--;
                $this->skipString($string);

                continue;
            }

            $this->addString($string);
        }

        if (count($this->strings) == 1) {
            return reset($this->strings);
        }

        return $this->strings;
    }

    public function useConfig(StringConfig $config): void
    {
        $this->config = $config;
    }

    public function getConfig(): StringConfig
    {
        return $this->config;
    }

    protected function addString(string $string): void
    {
        $this->strings[] = $string;
    }

    protected function getStrings(): array
    {
        return $this->strings;
    }

    protected function shouldSkip(string $string): bool
    {
        $shouldSkip = false;

        if ($this->config->isUnique()) {
            $shouldSkip = in_array($string, $this->getStrings());
        }

        if ($this->config->hasSkipCallback()) {
            $shouldSkip = $shouldSkip || $this->config->getSkipCallback()($string);
        }

        return $shouldSkip;
    }

    protected function reset(): void
    {
        $this->strings = [];
        $this->skipped = [];
    }

    protected function skipString(string $string): void
    {
        if (! in_array($string, $this->skipped)) {
            $this->skipped[] = $string;
        }
    }
}


src/StringConfig.php
<?php

declare(strict_types=1);

namespace Stfn\RandomString;

use Closure;

class StringConfig
{
    const CHARSET_UPPERCASE = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ';

    const CHARSET_LOWERCASE = 'abcdefghijklmnopqrstuvwxyz';

    const CHARSET_NUMERIC = '0123456789';

    protected int $length;

    protected int $count = 1;

    protected string $charset;

    protected string $prefix = '';

    protected string $suffix = '';

    protected bool $unique = false;

    protected Closure|null $skipCallback = null;

    public function __construct($length = 16)
    {
        $this->length = $length;
        $this->charset = self::CHARSET_LOWERCASE.self::CHARSET_UPPERCASE.self::CHARSET_NUMERIC;
    }

    public static function make($length = 16): self
    {
        return new self($length);
    }

    public static function fromArray(array $array): self
    {
        $object = new self();

        $allowedProperties = ['count', 'prefix', 'suffix', 'length', 'charset', 'unique'];

        foreach ($array as $property => $value) {
            if (property_exists($object, $property) && in_array($property, $allowedProperties)) {
                $object->{$property} = $value;
            }
        }

        return $object;
    }

    public function validate(): void
    {
        if (empty($this->charset)) {
            throw InvalidStringConfigException::invalidCharset();
        }

        if ($this->length < 1) {
            throw InvalidStringConfigException::propertyNotPositiveNumber('length');
        }

        if ($this->count < 1) {
            throw InvalidStringConfigException::propertyNotPositiveNumber('count');
        }
    }

    public function charset(string $charset): self
    {
        $this->charset = $charset;

        return $this;
    }

    public function upperCaseOnly(): self
    {
        $this->charset = self::CHARSET_UPPERCASE;

        return $this;
    }

    public function lowerCaseOnly(): self
    {
        $this->charset = self::CHARSET_LOWERCASE;

        return $this;
    }

    public function numbersOnly(): self
    {
        $this->charset = self::CHARSET_NUMERIC;

        return $this;
    }

    public function length(int $length): self
    {
        $this->length = $length;

        return $this;
    }

    public function count(int $count): self
    {
        $this->count = $count;

        return $this;
    }

    public function skip(callable $callback): self
    {
        $this->skipCallback = $callback;

        return $this;
    }

    public function prefix(string $prefix): self
    {
        $this->prefix = $prefix;

        return $this;
    }

    public function suffix(string $suffix): self
    {
        $this->suffix = $suffix;

        return $this;
    }

    public function unique(): self
    {
        $this->unique = true;

        return $this;
    }

    public function notUnique(): self
    {
        $this->unique = false;

        return $this;
    }

    public function hasSkipCallback(): bool
    {
        return is_callable($this->skipCallback);
    }

    public function getSkipCallback(): Closure|null
    {
        return $this->skipCallback;
    }

    public function getLength(): int
    {
        return $this->length;
    }

    public function getCharset(): string
    {
        return $this->charset;
    }

    public function getPrefix(): string
    {
        return $this->prefix;
    }

    public function getSuffix(): string
    {
        return $this->suffix;
    }

    public function isUnique(): bool
    {
        return $this->unique;
    }

    public function getCount(): int
    {
        return $this->count;
    }
}
